design/design-doc-developer-experience-tooling.md

```diff
@@ -0,0 +1,1146 @@
+  compositions and operations that consume them, and to test them together,
+  avoiding the need to embed source code in YAML manifests.
+* Support functions built in any language (including general purpose languages
+  and templating/configuration languages) and using any SDK or framework.
+* Enable authors to take advantage of IDE features such as tab completion and
+  syntax highlighting when building functions.
+* Design with extensibility in mind, such that we can enable users to plug in
+  their own implementations of individual parts of the DevEx in the future
+  without an architectural overhaul.
+
+## Prior Art and Related Work
+
+This design is based on the [Upbound developer experience], which has already
+been implemented as a proprietary tool. Upbound intends to contribute code from
+their proprietary tooling to the Crossplane community to implement this design.
+
+This design includes a section on testing. A number of tools have been built or
+proposed for testing Crossplane configurations, and may be integrated with this
+design. Examples include [xprin] and the proposed [`crossplane beta test`] tool.
+
+[function-pythonic] offers a Python-based developer experience for building
+composition functions, and could be integrated into this design as a builder
+(see the section on functions below).
+
+## Proposal
+
+### Overview
+
+> [!NOTE]
+> This document proposes a tree of commands for the `crossplane` CLI. These
+> commands will likely start out as `beta` commands, but are written in this
+> design without the `beta` prefix to demonstrate the ultimate future state. For
+> example, `crossplane project build` will initially be `crossplane beta project
+> build`.
+
+A project will have a directory layout similar to the following:
+
+```text
+project.yaml
+apis
+├── cluster
+│───├── definition.yaml
+│───├── composition.yaml
+examples
+├── cluster
+├───├── xr.yaml
+functions
+├── compose-cluster
+├───├── main.k
+├───├── helpers.k
+├── propagate-status
+├───├── go.mod
+├───├── go.sum
+├───├── main.go
+├── recycle-nodes
+├───├── main.py
+operations
+├── recycle-nodes
+│───├── operation.yaml
+tests
+├── e2etest-cluster-api
+├───├── test.yaml.gotmpl
+├── test-cluster-api
+├───├── main.py
+├── test-recycle-nodes
+├───├── go.mod
+├───├── go.sum
+├───├── main.go
+```
+
+This example project contains one composite type (`cluster`) supported by two
+functions (`compose-cluster` and `propagate-status`). The `compose-cluster`
+function is built in KCL, while `propagate-status` is built in Go. The project
+also contains one operation (`recycle-nodes`) that runs an operation function of
+the same name, built in Python. The `tests` directory contains tests for the
+composition and operation. We call the functions built as part of a project
+"embedded functions", since they sit alongside configuration rather than in
+their own repositories.
+
+The `project.yaml` file contains metadata and configuration for the project. It
```

Comment:
Would `crossplane-project.yaml` be too verbose? I'm thinking about e.g. LSPs that use root file patterns to figure out what type of project a directory is (e.g. `pyproject.toml` and `requirements.txt` for Python, `go.mod` for Go). `project.yaml` is quite generic.

Reply:
Updated to `crossplane-project.yaml`. It's a little long, but I agree it's nice to be more specific and I don't have a better idea.